src/test/ui/traits/bad-trait-error-span-in-call.stderr

```diff
@@ -0,0 +1,241 @@
+   |
+LL | fn example<Q: T3>(q: Q) {
+   |             ++++
+
+error[E0277]: the trait bound `Q: T3` is not satisfied
+  --> $DIR/bad-trait-error-span-in-call.rs:96:27
+   |
+LL |     want(Wrapper { value: AliasBurrito { spiciness: q, filling: q } });
+   |     ----                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `T3` is not implemented for `Q`
+   |     |
+   |     required by a bound introduced by this call
+   |
+note: required for `GenericBurrito<Q, Q>` to implement `T2`
+  --> $DIR/bad-trait-error-span-in-call.rs:47:16
+   |
+LL | impl<X, Y: T3> T2 for GenericBurrito<X, Y> {}
+   |                ^^     ^^^^^^^^^^^^^^^^^^^^
+note: required for `Wrapper<GenericBurrito<Q, Q>>` to implement `T1`
+  --> $DIR/bad-trait-error-span-in-call.rs:16:13
+   |
+LL | impl<B: T2> T1 for Wrapper<B> {}
+   |             ^^     ^^^^^^^^^^
+note: required by a bound in `want`
+  --> $DIR/bad-trait-error-span-in-call.rs:52:12
+   |
+LL | fn want<V: T1>(_x: V) {}
+   |            ^^ required by this bound in `want`
+help: consider restricting type parameter `Q`
+   |
+LL | fn example<Q: T3>(q: Q) {
+   |             ++++
+
+error[E0277]: the trait bound `Q: T1` is not satisfied
+  --> $DIR/bad-trait-error-span-in-call.rs:99:19
+   |
+LL |     want(Two { a: Two { a: (), b: q }, b: () });
+   |     ----          ^^^^^^^^^^^^^^^^^^^ the trait `T1` is not implemented for `Q`
+   |     |
+   |     required by a bound introduced by this call
+   |
+note: required for `Two<Two<(), Q>, ()>` to implement `T1`
+  --> $DIR/bad-trait-error-span-in-call.rs:65:19
+   |
+LL | impl<X, Y: T1, Z> T1 for Two<Two<X, Y>, Z> {}
+   |                   ^^     ^^^^^^^^^^^^^^^^^
+note: required by a bound in `want`
+  --> $DIR/bad-trait-error-span-in-call.rs:52:12
+   |
+LL | fn want<V: T1>(_x: V) {}
+   |            ^^ required by this bound in `want`
+help: consider restricting type parameter `Q`
+   |
+LL | fn example<Q: T1>(q: Q) {
+   |             ++++
+
+error[E0277]: the trait bound `Q: T1` is not satisfied
+  --> $DIR/bad-trait-error-span-in-call.rs:105:27
+   |
+LL |     want(
+   |     ---- required by a bound introduced by this call
+LL |         Two { a: Two { a: (), b: Two { a: Two { a: (), b: q }, b: () } }, b: () },
+   |                           ^^ the trait `T1` is not implemented for `Q`
+   |
+note: required for `Two<Two<(), Q>, ()>` to implement `T1`
+  --> $DIR/bad-trait-error-span-in-call.rs:65:19
+   |
+LL | impl<X, Y: T1, Z> T1 for Two<Two<X, Y>, Z> {}
+   |                   ^^     ^^^^^^^^^^^^^^^^^
+   = note: 1 redundant requirement hidden
+   = note: required for `Two<Two<(), Two<Two<(), Q>, ()>>, ()>` to implement `T1`
+note: required by a bound in `want`
+  --> $DIR/bad-trait-error-span-in-call.rs:52:12
+   |
+LL | fn want<V: T1>(_x: V) {}
+   |            ^^ required by this bound in `want`
+help: consider restricting type parameter `Q`
+   |
+LL | fn example<Q: T1>(q: Q) {
+   |             ++++
```

Comment:
This is a new test, which shows a bug in the current logic. I have some idea of how to address this.

You'll notice that the main span above is pointing at the `a: ()`, when it should point at `b: q`.

The problem here is that the impl *actually* should cause it to go `.a.b`, but it's currently only looking at the _first_ field and drilling that way. Basically, it needs to "drill into" the derived `impl` and make equivalent steps in the `expr` to be valid.

So in particular, if we have a constraint like `Y: Q1` for an `impl` with the `Self` type `Two<Two<X, Y>, Z>`, we need to "drill to the `Y`", wherever that might be. If we can get there *exactly*, then we recurse further. But if we get stuck at some point, we need to stop drilling.

---

So the updated strategy needs to add an extra recursive step: drilling to `Y` inside of `Two<Two<X, Y>, Z>` while simultaneously drilling `.a.b` inside of `expr`.